File: server/src/main/java/br/com/diegosilva/automation/dto/IOTMessage.java
package br.com.diegosilva.automation.dto;

import br.com.diegosilva.automation.CborSerializable;
import com.fasterxml.jackson.annotation.JsonCreator;

import java.util.Arrays;
import java.util.Map;
import java.util.stream.Collectors;

public class IOTMessage implements CborSerializable {

    public String id;
    public String sensor;
    public String value;

    @JsonCreator
    public IOTMessage() {

    }

    @JsonCreator
    public IOTMessage(String id, String sensor, String value) {
        this.id = id;
        this.sensor = sensor;
        this.value = value;
    }

    public static IOTMessage decode(String message) {
        try {
            Map<String, String> map = Arrays.stream(message.split(","))
                    .collect(Collectors.toMap(value -> value.split(":")[0], value -> value.split(":")[1]));
            IOTMessage retorno = new IOTMessage();

            if (map.containsKey("id"))
                retorno.id = map.get("id");
            else
                return null;

            if (map.containsKey("sen"))
                retorno.sensor = map.get("sen");
            else
                return null;

            if (map.containsKey("val"))
                retorno.value = map.get("val");
            else
                return null;

            return retorno;
        } catch (Exception e) {
            return null;
        }
    }

    public String encode() {
        return "id:" + this.id + "," + "sen:" + this.sensor + "," + "val:" + value;
    }

    @Override
    public String toString() {
        return "IOTMessage{" +
                "id='" + id + '\'' +
                ", sensor='" + sensor + '\'' +
                ", value='" + value + '\'' +
                '}';
    }
}

File: server/src/main/scala/br/com/diegosilva/home/dto/IOTMessage.scala
package br.com.diegosilva.home.dto


object IOTMessage {

  def decode(message: String): Option[IOTMessage] = {

    val arr = message.split(",")

    if (arr.length == 2) {
      val map: Map[String, String] = arr
        .map(value => value.split(":")(0) -> value.split(":")(1)).toMap
      val retorno = IOTMessage()
      if (map.contains("id"))
        retorno.id = map.getOrElse("id", "")
      else
        return None

      if (map.contains("sen"))
        retorno.sensor = map.getOrElse("sen", "")
      else
        return None

      if (map.contains("val"))
        retorno.value = map.getOrElse("val", "")
      else
        return None

      Some(retorno)
    } else {
      return None
    }
  }
}


case class IOTMessage(var id: String = "", var sensor: String = "", var value: String = "") {

  def encode: String = "id:" + this.id + "," + "sen:" + this.sensor + "," + "val:" + value

  override def toString: String = "IOTMessage{" + "id='" + id + '\'' + ", sensor='" + sensor + '\'' + ", value='" + value + '\'' + '}'
}
